File: CPP/3-Behavioral/Chain of Responsibility/ChainOfResponsibility.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>
#include <ctime>

class EventHandler
{
    EventHandler *_next; // "_next" pointer in the base class

public:
    EventHandler()
    {
        _next = 0;
    }

    void setNext(EventHandler *n)
    {
        _next = n;
    }

    void add(EventHandler *n)
    {
        if (_next)
        {
            _next->add(n);
        }
        else
        {
            _next = n;
        }
    }

    // The "chain" method in the base class
    // always delegates to the _next obj
    virtual void handle(int i)
    {
        _next->handle(i);
    }
};

class MouseEventHandler: public EventHandler
{
public:
    /*virtual*/ void handle(int i)
    {
        if (rand() % 3)
        {
            // Don't handle requests 3 times out of 4
            std::cout << "MouseEvent passed " << i << "  ";
            EventHandler::handle(i); // Delegate to the base class
        }
        else
        {
            std::cout << "MouseEvent handled " << i << "  ";
        }
    }
};

class KeyEventHandler: public EventHandler
{
public:
    /*virtual*/ void handle(int i)
    {
        if (rand() % 3)
        {
            std::cout << "KeyEvent passed " << i << "  ";
            EventHandler::handle(i);
        }
        else
        {
            std::cout << "KeyEvent handled " << i << "  ";
        }
    }
};

class InteractionEventHandler: public EventHandler
{
public:
    /*virtual*/ void handle(int i)
    {
        if (rand() % 3)
        {
            std::cout << "InteractionEvent passed " << i << "  ";
            EventHandler::handle(i);
        }
        else
        {
            std::cout << "InteractionEvent handled " << i << "  ";
        }
    }
};

int main()
{
    srand(time(0));
    MouseEventHandler mainUI;
    KeyEventHandler textField;
    InteractionEventHandler textArea;

    mainUI.add(&textField);
    mainUI.add(&textArea);
    textArea.setNext(&mainUI);

    for (int i = 1; i < 10; ++i)
    {
        mainUI.handle(i);
        std::cout << std::endl;
    }

    return 0;
}



//-----------------------------------------------------------
/*
Output:

MouseEvent passed 1  KeyEvent passed 1  InteractionEvent handled 1
MouseEvent passed 2  KeyEvent handled 2
MouseEvent passed 3  KeyEvent passed 3  InteractionEvent handled 3
MouseEvent handled 4
MouseEvent passed 5  KeyEvent passed 5  InteractionEvent passed 5  MouseEvent handled 5
MouseEvent handled 6
MouseEvent handled 7
MouseEvent passed 8  KeyEvent handled 8
MouseEvent passed 9  KeyEvent handled 9

*/

File: Python/3-Behavioural/ChainOfResponsibility/ChainOfResponsibility.py
#!/usr/bin/env python3


class EventHandler:
    """
    Base class of the event's handlers. It defines the handler's API
    """
    def __init__(self):
        self._next = None

    def set_next(self, handler):
        self._next = handler

    def add(self, handler):
        """
        It adds a given handler to the chain assigning it to the _next property
        :param handler: the handler to put in chain
        :return: None
        """
        if self._next is None:
            self._next = handler
        else:
            self._next.add(handler)

    def handle(self, ev):
        """
        It allows to handle the given event
        :param ev: the event to handle
        :return: None
        """
        if self._next is None:
            print("Event %d not handled!!", ev)
        else:
            self._next.handle(ev)


class MouseEventHandler(EventHandler):
    def handle(self, i):
        # Let's decide which is the rule for handling the Mouse events
        if 3 <= i < 5:
            print("MouseEvent handled %d\n" % i)
        else:
            print("MouseEvent passed")
            EventHandler.handle(self, i)


class KeyEventHandler(EventHandler):
    def handle(self, i):
        # Let's decide which is the rule for handling the Key events
        if i >= 5:
            print("KeyEventHandler handled %d\n" % i)
        else:
            print("KeyEventHandler passed")
            EventHandler.handle(self, i)


class InteractionEventHandler(EventHandler):
    def handle(self, i):
        # Let's decide which is the rule for handling the Interaction events
        if i <= 2:
            print("InteractionEventHandler handled %d\n" % i)
        else:
            print("InteractionEventHandler passed")
            EventHandler.handle(self, i)


def main():
    # Create some handlers
    main_ui = MouseEventHandler()
    text_field = KeyEventHandler()
    text_area = InteractionEventHandler()

    # ... and then chain them together
    main_ui.add(text_field)
    main_ui.add(text_area)
    # the next call will create a loop with the first event handler
    # Pay attention to be sure that all the handlers are able to
    # handle the events, otherwise you'll get a stack overflow.
    text_area.set_next(main_ui)

    # Now make some calls to request the event's handling
    for i in range(1, 10):
        main_ui.handle(i)


if __name__ == '__main__':
    main()


"""
Output:

MouseEvent passed
KeyEventHandler passed
InteractionEventHandler handled 1

MouseEvent passed
KeyEventHandler passed
InteractionEventHandler handled 2

MouseEvent handled 3

MouseEvent handled 4

MouseEvent passed
KeyEventHandler handled 5

MouseEvent passed
KeyEventHandler handled 6

MouseEvent passed
KeyEventHandler handled 7

MouseEvent passed
KeyEventHandler handled 8

MouseEvent passed
KeyEventHandler handled 9

"""
